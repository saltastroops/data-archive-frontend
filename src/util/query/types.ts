/**
 * An interface marking a where condition.
 */
/* tslint:disable:no-empty-interface */
export interface IWhereCondition {}
/* tslint:enable:no-empty-interface */

/**
 * An interface describing a target position.
 *
 * Properties:
 * -----------
 * declinations:
 *     An array of 0, 1 or 2 declinations, in degrees. If there are two
 *     declinations, they are taken to be the ends of an interval.
 * rightAscensions:
 *     An array of 0, 1 or 2 right ascensions, in degrees. If there are two
 *     right ascensions, they are taken to be the ends of an interval.
 * searchConeRadius:
 *     The radius for a cone search, in degrees.
 */
export interface ITargetPosition {
  rightAscensions: number[];
  declinations: number[];
  searchConeRadius: number;
}

/**
 * An interface describing the arguments for the WITHIN_RADIUS operator.
 *
 * Properties:
 * -----------
 * declination:
 *     The declination of the centre position, in degrees.
 * declinationColumn:
 *     The name of the declination column.
 * radius:
 *     The radius of the search cone.
 * rightAscension:
 *     The right ascension of the centre position, in degrees.
 * rightAscensionColumn:
 *     The name of the right ascension column.
 */
export interface IWithinRadiusArguments {
  declination: number;
  declinationColumn: string;
  radius: number;
  rightAscension: number;
  rightAscensionColumn: string;
}

// Interfaces describing the state of the search form

/**
 * An interface describing the React state of the observation query form.
 *
 * This state is used for populating all search form fields (and error messages)
 * and is parsed into the GraphQL query for searching observations.
 *
 * Properties:
 * -----------
 * general:
 *     General information about the observation.
 * target:
 *     Target details.
 * telescope:
 *     Telescope (and instrument) details.
 */
export interface IObservationQueryParameters {
  general: IGeneral;
  target: ITarget;
  telescope: ITelescope;
}

/**
 * An interface for query parameters related to general information.
 *
 * Properties:
 * -----------
 * errors:
 *     Error messages.
 * observationNight:
 *     Date of the night when the observation was taken. This
 *     refers to the date whe the night started.
 * proposalCode:
 *     Identifier of the proposal to which the observation belongs.
 * proposalTitle:
 *     Title of the proposal to which the observation belongs.
 *
 */
export interface IGeneral {
  errors: {
    observationNight?: string;
  };
  observationNight?: string;
  principalInvestigator?: string;
  proposalCode?: string;
}

/**
 * An interface for query parameters related to target details.
 *
 * Properties:
 * -----------
 * declination:
 *     Declination.
 * errors:
 *     Errors.
 * name:
 *     Target name.
 * resolver:
 *     Resolver to use for resolving the target name to a position.
 * rightAscension:
 *     Right ascension.
 * searchConeRadius:
 *     Radius of the search cone radius, in the units given by
 *     searchConeRadiusUnits.
 * searchConeRadiusUnits:
 *     Units of the search cone radius given by searchConeRadius.
 *
 */
export interface ITarget {
  declination?: string;
  errors: {
    declination?: string;
    resolver?: string;
    rightAscension?: string;
    searchConeRadius?: string;
    searchConeRadiusUnits?: string;
  };
  name?: string;
  resolver?: string;
  rightAscension?: string;
  searchConeRadius?: string;
  searchConeRadiusUnits?: string;
}

// TELESCOPES

/**
 * An interface for query parameters related to a telescope.
 *
 * This serves as a placeholder for more concrete interfaces such as ISALT.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface ITelescope {
  name: "Lesedi" | "SALT" | "1.9 m";
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 * name:
 *     The string "SALT".
 */
export interface ISALT extends ITelescope {
  instrument?: IInstrument;
  name: "SALT";
}

/**
 * An interface for query parameters related to Lesedi.
 *
 * Properties:
 * -----------
 * instrument:
 *     An instrument used on Lesedi.
 * name:
 *     The string "Lesedi".
 */
export interface ILesedi extends ITelescope {
  instrument?: IInstrument;
  name: "Lesedi";
}

/**
 * An interface for query parameters related to the 1.9 m Telescope.
 *
 * Properties:
 * -----------
 * instrument:
 *     An instrument used on the 1./9 m Telescope.
 * name:
 *     The string "1.9 m".
 */
export interface IOneDotNineMetre extends ITelescope {
  instrument?: IInstrument;
  name: "1.9 m";
}

// INSTRUMENTS

/**
 * An interface for query parameters related to an instrument.
 *
 * This serves as a placeholder for more concrete interfaces such as ISalticam.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface IInstrument {
  name: string;
}

/**
 * An interface for query parameters related to Salticam.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "Salticam".
 */
export interface ISalticam extends IInstrument {
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "Salticam";
}

/**
 * An interface for query parameters related to RSS.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "RSS".
 */
export interface IRSS extends IInstrument {
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "RSS";
}

/**
 * An interface for query parameters related to HRS.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * mode:
 *     Resolution mode.
 * name:
 *     The string "HRS".
 */
export interface IHRS extends IInstrument {
  errors: {
    exposureTime?: string;
    mode?: string;
  };
  exposureTime?: string;
  mode?:
    | "Low Resolution"
    | "Medium Resolution"
    | "High Resolution"
    | "High Stability";
  name: "HRS";
}

/**
 * An interface for query parameters related to HIPPO.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "HIPPO".
 */
export interface IHIPPO extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "HIPPO";
}

/**
 * An interface for query parameters related to SHOC.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "SHOC".
 */
export interface ISHOC extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "SHOC";
}
