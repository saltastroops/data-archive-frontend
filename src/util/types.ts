/**
 * Auth provider types.
 */
export type AuthProvider = "SSDA" | "SDB";

export interface IUser {
  authProvider: AuthProvider;
  familyName: string;
  givenName: string;
  isAdmin: boolean;
}
