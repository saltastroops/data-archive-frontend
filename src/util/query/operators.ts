import { IWhereCondition, IWithinRadiusArguments } from "./types";

type DATABASE_VALUE = string | number | boolean;

export function and(conditions: IWhereCondition[]): IWhereCondition {
  return {
    AND: conditions
  };
}

export function or(conditions: IWhereCondition[]): IWhereCondition {
  return {
    OR: conditions
  };
}

export function not(condition: IWhereCondition): IWhereCondition {
  return {
    NOT: condition
  };
}

export function equals(column: string, value: DATABASE_VALUE): IWhereCondition {
  return {
    EQUALS: { column, value }
  };
}

export function isNull(column: string) {
  return {
    IS_NULL: column
  };
}

export function lessThan(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    LESS_THAN: { column, value }
  };
}

export function greaterThan(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    GREATER_THAN: { column, value }
  };
}

export function lessEqual(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    LESS_EQUAL: { column, value }
  };
}

export function greaterEqual(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    GREATER_EQUAL: { column, value }
  };
}

export function isIn(
  column: string,
  values: DATABASE_VALUE[]
): IWhereCondition {
  return {
    IS_IN: { column, values }
  };
}

export function contains(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    CONTAINS: { column, value }
  };
}

export function startsWith(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    STARTS_WITH: { column, value }
  };
}

export function endsWith(
  column: string,
  value: DATABASE_VALUE
): IWhereCondition {
  return {
    ENDS_WITH: { column, value }
  };
}

export function withinRadius(args: IWithinRadiusArguments): IWhereCondition {
  return {
    WITHIN_RADIUS: args
  };
}
