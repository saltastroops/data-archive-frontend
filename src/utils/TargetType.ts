export type TargetType = string;

export const GALAXY = "10";

export const ISM = "20";

export const SOLAR_SYSTEM_BODY = "30";

export const STAR = "40";
