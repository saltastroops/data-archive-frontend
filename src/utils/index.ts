export const toTitleCase = (s: string) => {
  s = s.replace(/_/g, " ");
  return s.replace(/\w\S*/g, (txt: string) => {
    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
  });
};
