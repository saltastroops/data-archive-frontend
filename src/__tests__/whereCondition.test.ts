import { ISALT, ITarget } from "../util/query/types";
import {
  bvitWhereCondition,
  generalWhereCondition,
  hrsWhereCondition,
  rssWhereCondition,
  salticamWhereCondition,
  saltWhereCondition,
  targetWhereCondition,
  telescopeWhereCondition,
  whereCondition
} from "../util/query/whereCondition";
import {
  HRSMode,
  IGeneral,
  IHRS,
  IObservationQueryParameters,
  IRSS,
  ISalticam,
  RSSFabryPerotMode,
  RSSGrating,
  RSSInstrumentMode,
  RSSPolarimetryMode
} from "../utils/ObservationQueryParameters";
import { TargetType } from "../utils/TargetType";

describe("whereCondition", () => {
  it("should map query parameters correctly", () => {
    const general: IGeneral = {
      calibrations: new Set(),
      errors: {},
      observationNight: "2019-02-19",
      principalInvestigator: "John",
      proposalCode: "SCI"
    };
    const target: ITarget = {
      declination: "-42",
      errors: {},
      rightAscension: "12.34",
      searchConeRadius: "0.01",
      searchConeRadiusUnits: "degrees"
    };
    const rss: IRSS = {
      detectorMode: "Normal",
      errors: {},
      name: "RSS"
    };
    const salt: ISALT = {
      instrument: rss,
      name: "SALT"
    };
    const queryParameters: IObservationQueryParameters = {
      general,
      target,
      telescope: salt
    };
    const expected = {
      AND: [
        generalWhereCondition(general),
        targetWhereCondition(target),
        telescopeWhereCondition(salt)
      ]
    };
    expect(whereCondition(queryParameters)).toEqual(expected);
  });
});

describe("generalWhereCondition", () => {
  it("should raise an error for an invalid observation night format", () => {
    const f = () =>
      generalWhereCondition({
        errors: {},
        observationNight: "invaliddate",
        calibrations: new Set()
      });
    expect(f).toThrow(/invaliddate.*valid/);
  });

  it("should raise an error for an invalid observation night", () => {
    const f = () =>
      generalWhereCondition({
        errors: {},
        observationNight: "2019-02-29",
        calibrations: new Set()
      });
    expect(f).toThrow(/2019-02-29.*valid/);
  });

  it("should map an observation night to a greater equal and less equal condition spanning one night", () => {
    expect(
      generalWhereCondition({
        errors: {},
        observationNight: "2019-02-17",
        calibrations: new Set()
      })
    ).toMatchSnapshot();
  });

  it("should map a Principal Investigator to a contains condition", () => {
    expect(
      generalWhereCondition({
        errors: {},
        principalInvestigator: "Doe",
        calibrations: new Set()
      })
    ).toMatchSnapshot();
  });

  it("should map a proposal code to a contains condition", () => {
    expect(
      generalWhereCondition({
        errors: {},
        proposalCode: "2-SCI",
        calibrations: new Set()
      })
    ).toMatchSnapshot();
  });

  it("should map input data correctly", () => {
    expect(
      generalWhereCondition({
        calibrations: new Set(),
        errors: {},
        observationNight: "2019-05-13",
        principalInvestigator: "Sipho",
        proposalCode: "2019-1-SCI-042"
      })
    ).toMatchSnapshot();
  });
});

describe("targetWhereCondition", () => {
  it("should raise an error for invalid target position input", () => {
    const f = () =>
      targetWhereCondition({ errors: {}, rightAscension: "cghtyuu" });
    expect(f).toThrow();
  });

  it("should map a single position to a within condition", () => {
    expect(
      targetWhereCondition({
        declination: "-5",
        errors: {},
        rightAscension: "17",
        searchConeRadius: "0.04",
        searchConeRadiusUnits: "degrees"
      })
    ).toMatchSnapshot();
  });

  it("should map right ascension ranges to less than and greater than conditions", () => {
    expect(
      targetWhereCondition({ errors: {}, rightAscension: "13" })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({
        declination: "12 .. 13",
        errors: {},
        rightAscension: "13"
      })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({ errors: {}, rightAscension: "4 ... 355" })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({ errors: {}, rightAscension: "355 ... 4" })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({
        declination: "-12 .. -11",
        errors: {},
        rightAscension: "4 ... 355"
      })
    ).toMatchSnapshot();
  });

  it("should map declination ranges to less than and greater than conditions", () => {
    expect(
      targetWhereCondition({ errors: {}, declination: "-0.5" })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({
        declination: "-0.5",
        errors: {},
        rightAscension: "5 .. 9"
      })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({ declination: "-10 .. 12", errors: {} })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({ declination: "12 .. -10", errors: {} })
    ).toMatchSnapshot();
    expect(
      targetWhereCondition({
        declination: "10 .. 12",
        errors: {},
        rightAscension: "115 .. 200"
      })
    ).toMatchSnapshot();
  });

  it("should map target types to starts with conditions", () => {
    expect(
      targetWhereCondition({ errors: {}, targetTypes: new Set<TargetType>([]) })
    );
    expect(
      targetWhereCondition({
        errors: {},
        targetTypes: new Set<TargetType>(["Galaxy"])
      })
    );
    expect(
      targetWhereCondition({
        errors: {},
        targetTypes: new Set<TargetType>(["Galaxy", "Star"])
      })
    );
    expect(
      targetWhereCondition({
        errors: {},
        targetTypes: new Set<TargetType>(["ISM", "Solar System Body", "Galaxy"])
      })
    );
  });
});

describe("telescopeWhereCondition", () => {
  it("should map SALT query parameters correctly", () => {
    const rss: IRSS = {
      detectorMode: "Normal",
      errors: {},
      name: "RSS"
    };
    const salt: ISALT = {
      instrument: rss,
      name: "SALT"
    };
    const expected = {
      AND: [saltWhereCondition(salt)]
    };
    expect(telescopeWhereCondition(salt)).toEqual(expected);
  });
});

describe("saltWhereCondition", () => {
  it("should map Salticam query parameters correctly", () => {
    const salticam: ISalticam = {
      detectorMode: "Slot Mode",
      errors: {},
      name: "Salticam"
    };
    const salt: ISALT = {
      instrument: salticam,
      name: "SALT"
    };
    const expected = {
      AND: [
        { EQUALS: { column: "Telescope.telescopeName", value: "SALT" } },
        salticamWhereCondition(salticam)
      ]
    };
    expect(saltWhereCondition(salt)).toEqual(expected);
  });

  it("should map RSS query parameters correctly", () => {
    const rss: IRSS = {
      detectorMode: "Slot Mode",
      errors: {},
      name: "RSS"
    };
    const salt: ISALT = {
      instrument: rss,
      name: "SALT"
    };
    const expected = {
      AND: [
        { EQUALS: { column: "Telescope.telescopeName", value: "SALT" } },
        rssWhereCondition(rss)
      ]
    };
    expect(saltWhereCondition(salt)).toEqual(expected);
  });

  it("should map HRS query parameters correctly", () => {
    const hrs: IHRS = {
      errors: {},
      mode: "Medium Resolution",
      name: "HRS"
    };
    const salt: ISALT = {
      instrument: hrs,
      name: "SALT"
    };
    const expected = {
      AND: [
        { EQUALS: { column: "Telescope.telescopeName", value: "SALT" } },
        hrsWhereCondition(hrs)
      ]
    };
    expect(saltWhereCondition(salt)).toEqual(expected);
  });
});

describe("salticamWhereCondition", () => {
  it("should map the detector mode correctly", () => {
    const detectorModes: any = [
      "Drift Scan",
      "Frame Transfer",
      "Normal",
      "Slot Mode"
    ];
    const expectedModes = [
      "DRIFT SCAN",
      "FRAME TRANSFER",
      "NORMAL",
      "SLOT MODE"
    ];
    detectorModes.forEach((_, index: number) => {
      const salticam: ISalticam = {
        errors: {},
        name: "Salticam",
        detectorMode: detectorModes[index]
      };
      const expected = {
        AND: [
          {
            EQUALS: { column: "Instrument.instrumentName", value: "Salticam" }
          },
          {
            EQUALS: {
              column: "Salticam.detectorMode",
              value: expectedModes[index]
            }
          }
        ]
      };
      expect(salticamWhereCondition(salticam)).toEqual(expected);
    });
  });

  it("should map the filter correctly", () => {
    const filters: any = ["R-S1", "340-35", "SR708-25", "Sv-S1"];
    filters.forEach((filter: any) => {
      const salticam: ISalticam = {
        errors: {},
        name: "Salticam",
        filter
      };
      const expected = {
        AND: [
          {
            EQUALS: { column: "Instrument.instrumentName", value: "Salticam" }
          },
          { EQUALS: { column: "Salticam.filterName", value: filter } }
        ]
      };
      expect(salticamWhereCondition(salticam)).toEqual(expected);
    });
  });

  it("should combine conditions with an AND", () => {
    const salticam: ISalticam = {
      errors: {},
      name: "Salticam",
      detectorMode: "Frame Transfer",
      filter: "R-S1"
    };
    const expected = {
      AND: [
        { EQUALS: { column: "Instrument.instrumentName", value: "Salticam" } },
        {
          EQUALS: { column: "Salticam.detectorMode", value: "FRAME TRANSFER" }
        },
        { EQUALS: { column: "Salticam.filterName", value: "R-S1" } }
      ]
    };
    expect(salticamWhereCondition(salticam)).toEqual(expected);
  });
});

describe("rssWhereCondition", () => {
  it("should map the detector mode correctly", () => {
    const detectorModes: any = [
      "Drift Scan",
      "Frame Transfer",
      "Normal",
      "Shuffle",
      "Slot Mode"
    ];
    const expectedModes = [
      "DRIFT SCAN",
      "FRAME TRANSFER",
      "NORMAL",
      "SHUFFLE",
      "SLOT MODE"
    ];
    detectorModes.forEach((_, index: number) => {
      const rss: IRSS = {
        errors: {},
        name: "RSS",
        detectorMode: detectorModes[index]
      };
      const expected = {
        AND: [
          { EQUALS: { column: "Instrument.instrumentName", value: "RSS" } },
          {
            EQUALS: { column: "RSS.detectorMode", value: expectedModes[index] }
          }
        ]
      };
      expect(rssWhereCondition(rss)).toEqual(expected);
    });
  });

  it("should map the RSS modes correctly", () => {
    const modeNames: any = new Set([
      "Spectroscopy",
      "MOS",
      "FP Polarimatry",
      "Polarimetric Imaging"
    ]);
    const rss: IRSS = {
      errors: {},
      name: "RSS",
      modes: {
        errors: {},
        names: modeNames
      }
    };
    expect(rssWhereCondition(rss)).toMatchSnapshot();
  });

  it("should map the grating correctly", () => {
    const gratings: RSSGrating[] = ["Open", "pg1300", "pg3000"];
    gratings.forEach(grating => {
      const rss: IRSS = {
        errors: {},
        name: "RSS",
        modes: {
          errors: {},
          grating,
          names: new Set<RSSInstrumentMode>(["Spectroscopy"])
        }
      };
      const expected = {
        AND: [
          { EQUALS: { column: "Instrument.instrumentName", value: "RSS" } },
          { IS_IN: { column: "RssMode.rssMode", values: ["Spectroscopy"] } },
          { EQUALS: { column: "RSS.grating", value: grating } }
        ]
      };
      expect(rssWhereCondition(rss)).toEqual(expected);
    });
  });

  it("should map the Fabry-Perot mode correctly", () => {
    const fabryPerotModes: RSSFabryPerotMode[] = ["HR", "LR", "MR", "TF"];
    fabryPerotModes.forEach(fabryPerotMode => {
      const rss: IRSS = {
        errors: {},
        name: "RSS",
        modes: {
          errors: {},
          fabryPerotMode,
          names: new Set<RSSInstrumentMode>(["Spectroscopy"])
        }
      };
      const expected = {
        AND: [
          { EQUALS: { column: "Instrument.instrumentName", value: "RSS" } },
          { IS_IN: { column: "RssMode.rssMode", values: ["Spectroscopy"] } },
          {
            EQUALS: {
              column: "RssFabryPerotMode.rssFabryPerotMode",
              value: fabryPerotMode
            }
          }
        ]
      };
      expect(rssWhereCondition(rss)).toEqual(expected);
    });
  });

  it("should map the polarimetry modes correctly", () => {
    const polarimetryModes = new Set<RSSPolarimetryMode>([
      "LINEAR",
      "CIRCULAR"
    ]);
    const rss: IRSS = {
      errors: {},
      name: "RSS",
      modes: {
        errors: {},
        names: new Set<RSSInstrumentMode>([
          "Spectropolarimetry",
          "Polarimetric imaging"
        ]),
        polarimetryModes
      }
    };
    expect(rssWhereCondition(rss)).toMatchSnapshot();
  });
});

describe("hrsWhereCondition", () => {
  it("should map the mode correctly", () => {
    const modes: HRSMode[] = [
      "LOW RESOLUTION",
      "HIGH RESOLUTION",
      "INT CAL FIBRE"
    ];
    modes.forEach(mode => {
      const hrs: IHRS = {
        name: "HRS",
        errors: {},
        mode
      };
      const expected = {
        AND: [
          { EQUALS: { column: "Instrument.instrumentName", value: "HRS" } },
          { EQUALS: { column: "HRS.observationMode", value: mode } }
        ]
      };
      expect(hrsWhereCondition(hrs)).toEqual(expected);
    });
  });
});

describe("bvitWhereCondition", () => {
  // TODO: Add unit tests
  it("should work", () => {
    // to be filled with life
  });
});
