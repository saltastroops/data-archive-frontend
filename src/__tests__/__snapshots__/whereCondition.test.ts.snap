// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generalWhereCondition should map a Principal Investigator to a contains condition 1`] = `
Object {
  "AND": Array [
    Object {
      "CONTAINS": Object {
        "column": "proposal.pi",
        "value": "Doe",
      },
    },
  ],
}
`;

exports[`generalWhereCondition should map a proposal code to a contains condition 1`] = `
Object {
  "AND": Array [
    Object {
      "CONTAINS": Object {
        "column": "proposal.proposal_code",
        "value": "2-SCI",
      },
    },
  ],
}
`;

exports[`generalWhereCondition should map an observation night to a greater equal and less equal condition spanning one night 1`] = `
Object {
  "AND": Array [
    Object {
      "EQUALS": Object {
        "column": "observation_time.night",
        "value": "2019-02-17",
      },
    },
  ],
}
`;

exports[`generalWhereCondition should map input data correctly 1`] = `
Object {
  "AND": Array [
    Object {
      "EQUALS": Object {
        "column": "observation_time.night",
        "value": "2019-05-13",
      },
    },
    Object {
      "CONTAINS": Object {
        "column": "proposal.pi",
        "value": "Sipho",
      },
    },
    Object {
      "CONTAINS": Object {
        "column": "proposal.proposal_code",
        "value": "2019-1-SCI-042",
      },
    },
  ],
}
`;

exports[`generalWhereCondition should map input data correctly 2`] = `
Object {
  "AND": Array [
    Object {
      "AND": Array [
        Object {
          "GREATER_EQUAL": Object {
            "column": "observation_time.night",
            "value": "2019-05-01",
          },
        },
        Object {
          "LESS_EQUAL": Object {
            "column": "observation_time.night",
            "value": "2019-05-13",
          },
        },
      ],
    },
    Object {
      "CONTAINS": Object {
        "column": "proposal.pi",
        "value": "Sipho",
      },
    },
    Object {
      "CONTAINS": Object {
        "column": "proposal.proposal_code",
        "value": "2019-1-SCI-042",
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map a single position to a within condition 1`] = `
Object {
  "AND": Array [
    Object {
      "WITHIN_RADIUS": Object {
        "declination": -5,
        "declinationColumn": "position.dec",
        "radius": 0.04,
        "rightAscension": 17,
        "rightAscensionColumn": "position.ra",
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map declination ranges to less than and greater than conditions 1`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": -0.55,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": -0.45,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map declination ranges to less than and greater than conditions 2`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.ra",
        "value": 5,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.ra",
        "value": 9,
      },
    },
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": -0.55,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": -0.45,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map declination ranges to less than and greater than conditions 3`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": -10,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": 12,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map declination ranges to less than and greater than conditions 4`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": -10,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": 12,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map declination ranges to less than and greater than conditions 5`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.ra",
        "value": 115,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.ra",
        "value": 200,
      },
    },
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": 10,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": 12,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map right ascension ranges to less than and greater than conditions 1`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.ra",
        "value": 12.95,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.ra",
        "value": 13.05,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map right ascension ranges to less than and greater than conditions 2`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.ra",
        "value": 12.95,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.ra",
        "value": 13.05,
      },
    },
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": 12,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": 13,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map right ascension ranges to less than and greater than conditions 3`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.ra",
        "value": 4,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.ra",
        "value": 355,
      },
    },
  ],
}
`;

exports[`targetWhereCondition should map right ascension ranges to less than and greater than conditions 4`] = `
Object {
  "AND": Array [
    Object {
      "OR": Array [
        Object {
          "AND": Array [
            Object {
              "GREATER_EQUAL": Object {
                "column": "position.ra",
                "value": 355,
              },
            },
            Object {
              "LESS_EQUAL": Object {
                "column": "position.ra",
                "value": 360,
              },
            },
          ],
        },
        Object {
          "AND": Array [
            Object {
              "GREATER_EQUAL": Object {
                "column": "position.ra",
                "value": 0,
              },
            },
            Object {
              "LESS_EQUAL": Object {
                "column": "position.ra",
                "value": 4,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`targetWhereCondition should map right ascension ranges to less than and greater than conditions 5`] = `
Object {
  "AND": Array [
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.ra",
        "value": 4,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.ra",
        "value": 355,
      },
    },
    Object {
      "GREATER_EQUAL": Object {
        "column": "position.dec",
        "value": -11,
      },
    },
    Object {
      "LESS_EQUAL": Object {
        "column": "position.dec",
        "value": -12,
      },
    },
  ],
}
`;

exports[`telescopeWhereCondition should map All correctly 1`] = `
Object {
  "AND": Array [],
}
`;

exports[`telescopeWhereCondition should map All together with other items correctly 1`] = `
Object {
  "AND": Array [],
}
`;

exports[`telescopeWhereCondition should map conditions correctly 1`] = `
Object {
  "AND": Array [
    Object {
      "IS_IN": Object {
        "column": "detector_mode.detector_mode",
        "values": Array [
          "Normal",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "filter.name",
        "values": Array [
          "Johnson U",
          "Cousins I",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "hrs_mode.hrs_mode",
        "values": Array [
          "High Stability",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "instrument.name",
        "values": Array [
          "RSS",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "instrument_mode.instrument_mode",
        "values": Array [
          "Imaging",
          "Spectroscopy",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "rss_fabry_perot_mode.fabry_perot_mode",
        "values": Array [
          "MR",
          "HR",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "rss_grating.grating",
        "values": Array [
          "pg0900",
          "pg1800",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "polarization_mode.name",
        "values": Array [
          "LINEAR",
        ],
      },
    },
    Object {
      "IS_IN": Object {
        "column": "telescope.name",
        "values": Array [
          "Lesedi",
          "SALT",
        ],
      },
    },
  ],
}
`;

exports[`telescopeWhereCondition should map empty arrays correctly 1`] = `
Object {
  "AND": Array [],
}
`;
